pgdog/src/backend/pool/dns_cache.rs

```diff
@@ -0,0 +1,362 @@
+            hostnames: Arc::new(RwLock::new(HashSet::new())),
+        }
+    }
+
+    /// Resolve hostname to socket address string.
+    pub async fn resolve(&self, hostname: &str) -> Result<IpAddr, Error> {
+        if let Some(ip) = self.get_cached_ip(hostname) {
+            return Ok(ip);
+        }
+
+        // Track hostname for future refreshes.
+        {
+            let mut hostnames = self.hostnames.write();
+            if hostnames.insert(hostname.to_string()) {
+                debug!("dns cache added hostname \"{}\"", hostname);
+            }
+        }
+
+        let ip = self.resolve_and_cache(hostname).await?;
+        Ok(ip)
+    }
+
+    /// Start the background refresh loop.
+    /// This spawns an infinite loop that runs until the process exits.
+    /// For testing, use short TTLs and allow the test runtime to handle shutdown.
+    pub fn start_refresh_loop(self: &Arc<Self>) {
+        let interval = Self::ttl();
+
+        let cache_ref = Arc::clone(self);
+        tokio::spawn(async move {
+            loop {
+                println!("-> I AM REFRESHING!!! {:?}", interval);
+                sleep(interval).await;
+                cache_ref.refresh_all_hostnames().await;
+            }
+        });
+    }
+}
+
+// -------------------------------------------------------------------------------------------------
+// ------ DnsCache :: Private methods --------------------------------------------------------------
+
+impl DnsCache {
+    /// Get the DNS refresh interval (short in tests, config-based otherwise).
+    fn ttl() -> Duration {
+        if cfg!(test) {
+            return Duration::from_millis(50);
+        }
+
+        config()
+            .config
+            .general
+            .dns_ttl()
+            .unwrap_or(Duration::from_secs(60)) // Should never happen in practice
+    }
+
+    /// Get IP address from cache only. Returns None if not cached or expired.
+    fn get_cached_ip(&self, hostname: &str) -> Option<IpAddr> {
+        if let Ok(ip) = hostname.parse::<IpAddr>() {
+            return Some(ip);
+        }
+
+        self.cache.read().get(hostname).copied()
+    }
+
+    /// Refresh all hostnames in the refresh list.
+    async fn refresh_all_hostnames(self: &Arc<Self>) {
+        let hostnames_to_refresh: Vec<String> = self.hostnames.read().iter().cloned().collect();
+
+        let tasks: Vec<_> = hostnames_to_refresh
+            .into_iter()
+            .map(|hostname| {
+                let cache_ref = Arc::clone(self);
+                tokio::spawn(async move {
+                    if let Err(e) = cache_ref.resolve_and_cache(&hostname).await {
+                        error!("failed to refresh DNS for {}: {}", hostname, e);
+                        println!("failed to refresh DNS for {}: {}", hostname, e);
+                    } else {
+                        println!("refreshed hostname: {}", hostname);
+                        debug!("refreshed hostname: {}", hostname);
```

Comment:
```suggestion
                        debug!("refreshed hostname \"{}\"", hostname);
```

To be consistent with other debug log.